services/brig/src/Brig/Run.hs

```diff
@@ -103,3 +124,51 @@ lookupRequestIdMiddleware :: (RequestId -> Wai.Application) -> Wai.Application
 lookupRequestIdMiddleware mkapp req cont = do
   let reqid = maybe def RequestId $ lookupRequestId req
   mkapp reqid req cont
+
+cleanExpiredPendingInvitations :: AppIO ()
+cleanExpiredPendingInvitations = do
+  safeForever "cleanExpiredPendingInvitations" $ do
+    today <- utctDay <$> (liftIO =<< view currentTime)
+    for_ [0 .. 9] $ \i ->
+      cleanUpDay (addDays (- i) today)
+    let d :: Int = 24 * 60 * 60
+    randomSecs <- liftIO (round <$> randomRIO @Double (0.5 * fromIntegral d, fromIntegral d))
+    threadDelay (randomSecs * 1_000_000)
+  where
+    cleanUpDay :: Day -> AppIO ()
+    cleanUpDay day = do
+      expiredEntries <- forTrackedExpirations day $ \exps ->
+        catMaybes
+          <$> ( for exps $ \(PendingActivationExpiration uid expiresAt tid) -> do
+                  isExpired <- (expiresAt <=) <$> (liftIO =<< view currentTime)
+                  if isExpired
+                    then do
+                      isPendingInvitation <- (Just PendingInvitation ==) <$> lookupStatus uid
+                      invExpired <- isNothing <$> Data.lookupInvitation tid (coerce uid)
+                      when (isPendingInvitation && invExpired) $ do
+                        API.deleteUserNoVerify uid
+                      pure (Just uid)
+                    else pure Nothing
+              )
+      unless (null expiredEntries) $
+        removeTrackedExpiration day expiredEntries
+
+    forTrackedExpirations :: Day -> ([PendingActivationExpiration] -> AppIO [UserId]) -> AppIO [UserId]
```

Comment:
It would be good for space usage if you could process one page fully (including deletions) before pulling the next one.  Not sure what cassandra has to say about this (@jschaul, again?  sorry..).  This function could look like this:

```suggestion
    forTrackedExpirations :: Day -> (Page PendingActivationExpiration -> AppIO ()) -> AppIO ()
```

And the conditional call to `removeTrackedExpiration(s)` above could be moved into the callback and run once per page.

If it turns out problematic to paginate through a select query while deleting stuff in the same table, you can just keep pulling the first page, and it'll always be the next one because you just deleted the previous one.  (Not sure that's any better on performance, but note that we're only worried about system resources here, not response time -- it's fine to wait for cassandra to process the deletions if that takes time.)

Reply:
I've implemented "always pulling the first" page. It will loop until no more entries are returned from the query.